src/stdlib_linalg.fypp

```diff
@@ -108,4 +201,197 @@ contains
         end do
       end function trace_${t1[0]}$${k1}$
     #:endfor
+
+
+    #:for k1, t1 in RCI_KINDS_TYPES
+      pure function is_square_${t1[0]}$${k1}$(A) result(res)
+        ${t1}$, intent(in) :: A(:,:)
+        logical :: res
+        integer :: A_shape(2)
+        A_shape = shape(A)
+        res = (A_shape(1) .eq. A_shape(2))
+      end function is_square_${t1[0]}$${k1}$
+    #:endfor
+
+
+    #:for k1, t1 in RCI_KINDS_TYPES
+      pure function is_diagonal_${t1[0]}$${k1}$(A) result(res)
+        ${t1}$, intent(in) :: A(:,:)
+        logical :: res
+        ${t1}$ :: zero
+        integer :: A_shape(2), m, n, o, i, j
+        zero = 0 !zero of relevant type
+        A_shape = shape(A)
+        m = A_shape(1)
+        n = A_shape(2)
+        do j = 1, n !loop over all columns
+            o = min(j-1,m) !index of row above diagonal (or last row)
+            do i = 1, o !loop over rows above diagonal
+                if (.not. (A(i,j) .eq. zero)) then
```

Comment:
Is there any difference between the following two checks?
```
(A(i,j) == zero)

! vs

(A(i,j) < tiny(A))
```
I'm just wondering if there is an advantage to either of them.

Reply:
In my reading [here](https://stackoverflow.com/questions/65657679/tinyx-intrinsic-function), and a very simple test, there does appear to be some difference.

`(A(i,j) == zero)` will be true when `A(i,j)` is zero or smaller in magnitude than the smallest subnormal number (underflowing, not representable with any precision, effectively zero). The smallest subnormal is accessed by `tiny(x)*epsilon(x)`, and subnormals are available so long as IEEE754 is supported (exceedingly likely).

Meanwhile, `A(i,j) < tiny(A)` may be true even if `A(i,j)` is subnormal (hence still representable and not strictly equal to zero under IEEE754). I think it is probably not a good idea to use this version of the check, since you are effectively not taking full advantage of IEEE754 (even though your matrix may be VERY nearly diagonal, etc.). This being said, I believe other contributors have expressed less desire to know whether their matrices were exactly diagonal (for example), and  more desire to know whether the matrix is nearly diagonal (off diagonals being less than some possibly user defined `tol`. Such an implementation would be a very simple extension to the methods given here, and would let users define `tol` instead of imposing `tol=tiny(A)` somewhat arbitrarily (in the scheme of floating point) on them.

Two versions which I believe are equivalent and utilize subnormal numbers:
```fypp
(A(i,j) == zero) !using my definition of zero variable in the code
!and
(abs(A(i,j)) < tiny(A)*epsilon(A)) !obviously precompute the right side
```

My final opinion (if I understood all of this correctly) would be to keep the code as written since:
- I believe it's slightly easier to understand (and requires less knowledge of floating point) while being equivalent to a version that more clearly acknowledges IEEE754
- it does not require an `abs` call (hence it may be slightly faster)

Reply:
Thanks for doing the research. I agree you should keep the code as already written.

We can discuss the `is_diagonal(A,tol)` (or perhaps is `is_nearly_diagonal` or `is_diagonally_dominant`) in  #482 first.